sdk/core/azure-core-http-okhttp/src/main/java/com/azure/core/http/okhttp/OkHttpAsyncHttpClient.java

```diff
@@ -26,18 +37,23 @@
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.time.Duration;
 import java.util.Objects;
 
 /**
  * HttpClient implementation for OkHttp.
  */
 class OkHttpAsyncHttpClient implements HttpClient {
-    final OkHttpClient httpClient;
-    //
+
     private static final Mono<okio.ByteString> EMPTY_BYTE_STRING_MONO = Mono.just(okio.ByteString.EMPTY);
+    private static final long BUFFERED_FLUX_REQUEST_THRESHOLD = 10 * 1024L;
```

Comment:
this is still under investigation.

Reply:
uploadfromfile

### 10240
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 195.47&pm;10.23 | 195.03&pm;7.10 |
| 10 | 1957.91&pm;52.51 | 1986.58&pm;31.22 |
| 30 | 4710.36&pm;53.01 | 5021.21&pm;65.17 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 209.71 | 203.33 |
| 10 | 2037.15 | 2042.68 |
| 30 | 4781.55 | 5114.66 |

### 51200
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 181.69&pm;11.40 | 179.23&pm;16.21 |
| 10 | 1803.34&pm;93.47 | 1801.17&pm;27.77 |
| 30 | 3828.50&pm;30.90 | 4012.91&pm;45.57 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 191.97 | 189.32 |
| 10 | 1891.49 | 1830.78 |
| 30 | 3872.67 | 4054.30 |

### 102400
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 160.66&pm;4.76 | 158.46&pm;5.03 |
| 10 | 1626.22&pm;36.66 | 1628.13&pm;56.89 |
| 30 | 2319.71&pm;0.24 | 2310.05&pm;25.08 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 166.36 | 166.58 |
| 10 | 1663.43 | 1714.62 |
| 30 | 2319.96 | 2320.04 |

### 512000
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 61.85&pm;2.85 | 59.69&pm;1.63 |
| 10 | 466.09&pm;1.24 | 466.03&pm;1.04 |
| 30 | 466.20&pm;1.25 | 466.44&pm;0.60 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 65.09 | 61.71 |
| 10 | 466.64 | 466.63 |
| 30 | 466.71 | 466.71 |

### 1048576
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 47.30&pm;1.29 | 42.87&pm;2.04 |
| 10 | 226.72&pm;2.29 | 227.48&pm;1.41 |
| 30 | 228.08&pm;0.01 | 228.04&pm;0.01 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 48.66 | 45.18 |
| 10 | 228.03 | 228.02 |
| 30 | 228.09 | 228.05 |

### 10485760
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 14.52&pm;0.32 | 11.42&pm;0.33 |
| 10 | 22.83&pm;0.00 | 22.82&pm;0.01 |
| 30 | 22.86&pm;0.01 | 22.84&pm;0.01 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> old |
|--|--|--|
| 1 | 14.99 | 11.80 |
| 10 | 22.83 | 22.82 |
| 30 | 22.87 | 22.86 |

Reply:
New benchmarks that has reduced thread hopping suggested by @anuchandy 

These benchmarks were run without the "buffering threshold". 

Legend
"new" - this PR with aggressive hopping to bounded elastic scheduler
"new-main-2" - hopping to bounded elastic conditionally - (i.e. not hopping in the scenario tested)
"old" - main without these changes.

uploadfromfile

### 10240
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 211.91&pm;5.72 | 213.78&pm;5.74 | 211.38&pm;3.73 |
| 10 | 2107.92&pm;30.42 | 2129.70&pm;29.46 | 2119.20&pm;21.90 |
| 30 | 4705.16&pm;47.74 | 4945.65&pm;47.22 | 5024.73&pm;88.92 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 221.22 | 222.57 | 218.11 |
| 10 | 2156.30 | 2169.62 | 2157.91 |
| 30 | 4748.49 | 5028.95 | 5175.12 |

### 51200
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 195.09&pm;5.05 | 199.69&pm;5.20 | 194.97&pm;4.49 |
| 10 | 1981.50&pm;58.39 | 2030.13&pm;50.48 | 1992.45&pm;41.23 |
| 30 | 3860.39&pm;12.07 | 4046.30&pm;28.12 | 4066.19&pm;28.91 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 200.93 | 208.93 | 203.66 |
| 10 | 2089.54 | 2091.41 | 2032.85 |
| 30 | 3876.65 | 4095.31 | 4113.15 |

### 102400
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 176.98&pm;3.43 | 180.62&pm;5.06 | 172.80&pm;3.31 |
| 10 | 1768.82&pm;33.42 | 1844.21&pm;14.87 | 1733.89&pm;29.77 |
| 30 | 2319.60&pm;0.17 | 2319.78&pm;0.25 | 2319.72&pm;0.22 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 182.03 | 188.14 | 177.44 |
| 10 | 1826.52 | 1874.59 | 1785.00 |
| 30 | 2319.82 | 2320.17 | 2320.00 |

### 512000
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 65.53&pm;1.38 | 66.44&pm;0.76 | 61.32&pm;1.15 |
| 10 | 466.58&pm;0.08 | 466.63&pm;0.01 | 466.57&pm;0.15 |
| 30 | 466.72&pm;0.01 | 466.70&pm;0.02 | 466.71&pm;0.03 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 67.61 | 67.56 | 63.24 |
| 10 | 466.65 | 466.65 | 466.64 |
| 30 | 466.73 | 466.72 | 466.74 |

### 1048576
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 49.58&pm;0.85 | 49.70&pm;0.69 | 45.30&pm;0.59 |
| 10 | 228.02&pm;0.02 | 228.01&pm;0.03 | 228.02&pm;0.00 |
| 30 | 228.07&pm;0.00 | 228.07&pm;0.01 | 228.06&pm;0.02 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 51.59 | 50.94 | 46.33 |
| 10 | 228.03 | 228.03 | 228.02 |
| 30 | 228.08 | 228.09 | 228.11 |

### 10485760
#### Avg throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 15.46&pm;0.24 | 13.89&pm;0.41 | 12.02&pm;0.18 |
| 10 | 22.83&pm;0.00 | 22.83&pm;0.00 | 22.82&pm;0.00 |
| 30 | 22.86&pm;0.00 | 22.86&pm;0.00 | 22.84&pm;0.01 |
#### Max throughput (ops/s)
| Concurrency | okhttp <br> new | okhttp <br> new-main-2 | okhttp <br> old |
|--|--|--|--|
| 1 | 15.88 | 14.52 | 12.30 |
| 10 | 22.83 | 22.83 | 22.82 |
| 30 | 22.87 | 22.87 | 22.86 |